src/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs

```diff
@@ -879,6 +1042,34 @@ public static MethodCallExpression Call(Type type, string methodName, Type[] typ
         ///<paramref name="instance" />.Type is not assignable to the declaring type of the method represented by <paramref name="method" />.-or-The number of elements in <paramref name="arguments" /> does not equal the number of parameters for the method represented by <paramref name="method" />.-or-One or more of the elements of <paramref name="arguments" /> is not assignable to the corresponding parameter for the method represented by <paramref name="method" />.</exception>
         public static MethodCallExpression Call(Expression instance, MethodInfo method, IEnumerable<Expression> arguments)
         {
+            IReadOnlyList<Expression> argumentList = arguments as IReadOnlyList<Expression>;
```

Comment:
I have entered an issue to track this.
https://github.com/dotnet/coreclr/issues/1592

We should have better understanding what particular scenarios to avoid (if perf is bad for a reason) or perhaps fix it.